### ISSUE TITLE
Pyright error for assign repr method

### ISSUE THREAD

**Issue by kumaraditya303** [microsoft/pyright#2532]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Pyright errors for assigning repr method to another function.

**To Reproduce**
Steps to reproduce the behavior.

**Expected behavior**
No error

**Screenshots or Code**
```python
# pyright: strict
from typing import Type, TypeVar

T = TypeVar("T")


def auto_repr(t: Type[T]) -> Type[T]:
    def __repr__(self: T) -> str:
        items = (
            f"{k}={getattr(self,k)!r}"
            for k, v in self.__class__.__dict__.items()
            if isinstance(v, property)
        )
        return f"<{self.__class__.__name__} {' '.join(items)}>"

    t.__repr__ = __repr__
    return t


@auto_repr
class A:
    @property
    def hello(self) -> str:
        return "Hello"

    @property
    def world(self) -> str:
        return "World"


print(repr(A()))

```
**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

```console
pyright 1.1.184
```

**Comment by erictraut** [comment]
Yes, this behavior is intended. A type checker will not allow you to simply overwrite a class's method from outside that class.

You can use `# type: ignore` if you want to ignore this error.

**Comment by kumaraditya303** [comment]
> Yes, this behavior is intended. A type checker will not allow you to simply overwrite a class's method from outside that class.
> 
> You can use `# type: ignore` if you want to ignore this error.

@erictraut If that's the case, then why doesn't this code generate the same error ?
```python
class A:
    def f(self)->None:pass
    def g(self)->None:pass

A.f = A.g
```

**Comment by erictraut** [comment]
That's a good point. I'll take another look.

**Comment by kumaraditya303** [comment]
It is would be nice feature if pyright would accept assigning methods afterwards to classes or any other object if it's signature matches and the method is already exists like `__repr__` which all objects have.

**Comment by kumaraditya303** [comment]
There is also a mypy issue for this feature https://github.com/python/mypy/issues/2427
which has quite a lot of upvotes, so it might be worth a try in pyright.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.186, which I just published. It will also be included in the next release of pylance.

**Comment by erictraut** [comment]
I've decided to undo this change, which means that this code will no longer type check without errors. The special-case code that I added to handle it breaks other more common cases such as [this](https://github.com/microsoft/pyright/issues/3361).

The problem here is that `Type[T]` must be made concrete to determine what `__repr__` refers to, and it is ambiguous in the Python type system what the concrete form of `Type[T]` is. It can either be "concretized" as `object` (i.e. the object class) or an instance of the `type` class. The evaluation of the expression `t.__repr__` depends on the interpretation.

The type evaluation of the expression `t.__repr__` depends on which interpretation to use for the concrete form. If it is replacing the `__repr__` method on the `type` class, then it is bound as part of the member access, and there is no `self` parameter associated with the bound method, and this leads to a type violation.
